File: product-order-service/src/main/java/com/example/productorderservice/product/ProductController.java
package com.example.productorderservice.product;

import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequiredArgsConstructor
@RequestMapping("/products")
public class ProductController {

    private final ProductService productService;

    @PostMapping
    public ResponseEntity<Void> addProduct(@RequestBody AddProductRequest request) {
        productService.addProduct(request);
        return ResponseEntity.status(HttpStatus.CREATED).build();
    }

    @GetMapping("/{productId}")
    public ResponseEntity<GetProductResponse> getProduct(@PathVariable final Long productId) {
        return ResponseEntity.ok(productService.getProduct(productId));
    }

    @PatchMapping("/{productId}")
    public ResponseEntity<Void> updateProduct(
            @PathVariable final Long productId,
            @RequestBody final UpdateProductRequest request) {
        productService.updateProduct(productId, request);
        return ResponseEntity.ok().build();
    }
}

File: distributed-transaction/saga-product/src/main/kotlin/org/example/product/controller/ProductController.kt
package org.example.product.controller

import org.example.product.application.ProductService
import org.example.product.application.RedisLockService
import org.example.product.controller.dto.ProductBuyRequest
import org.example.product.controller.dto.ProductBuyResponse
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class ProductController(
    private val productService: ProductService,
    private val redisLockService: RedisLockService,
) {

    @PostMapping("/products/buy")
    fun buy(
        @RequestBody request: ProductBuyRequest,
    ): ProductBuyResponse {
        val key = "product:orchestration:${request.requestId}"

        val acquiredLock = redisLockService.tryLock(key, request.requestId)

        if (!acquiredLock) {
            throw RuntimeException("락 획득에 실패하였습니다.")
        }

        try {
            val result = productService.buy(request.toCommand())
            return ProductBuyResponse(result.totalPrice)
        } finally {
            redisLockService.releaseLock(key)
        }
    }
}